### PR TITLE
Optimize PyArrow block shuffling with dynamic rebatched chunk interleaving

### PR DESCRIPTION
**Pull request by yuxin00j**
When streaming PyArrow datasets with shuffling enabled (`buffer_size > 0`), the `BufferShuffledExamplesIterable._iter_arrow()` method historically relied on a Python-level `list` to maintain the shuffle buffer, replacing elements one-by-one. In a multi-worker `DataLoader` environment, this per-row generator approach caused severe GIL contention, high generator state-transition overhead, and inefficient memory interleaving. 
This PR introduces **Amortized Native C++ Block Shuffling** for PyArrow streaming. It completely replaces the Python 1-row exchange loop with vectorized block shuffling:
- Incoming PyArrow tables are accumulated in memory until the combined size reaches `2 * self.buffer_size`.
- The buffered tables are concatenated via `pyarrow.concat_tables(tables)`.
- The entire block is shuffled in a single vectorized C++ call using `pyarrow.compute.take(buffer_table, indices)`.
- The first `current_len - self.buffer_size` rows are yielded immediately, while the remaining `self.buffer_size` rows are retained as the seed buffer for the next shuffle block.
By eliminating the 1-row `yield` loop, we drastically drop CPU overhead and unlock the full IO throughput potential of PyArrow when combined with `split_dataset_by_node` and multiple dataloader workers.
### Impact
We measured workloads using a high number of concurrent shards (`max_buffer_input_shards=10`) and combined the change in this PR and PR #8387:

```python
import datasets
from torch.utils.data import DataLoader
from datasets.distributed import split_dataset_by_node

ds = datasets.load_dataset(
    "parquet",
    data_files="gs://your-bucket/benchmark_data/shard_*.parquet",
    split="train",
    streaming=True
)

ds = ds.shuffle(seed=0, buffer_size=10000, max_buffer_input_shards=10)
ds = split_dataset_by_node(ds, rank=0, world_size=8)
ds = ds.with_format("torch")

loader = DataLoader(
    ds,
    batch_size=8,
    num_workers=4,
    prefetch_factor=2,
    persistent_workers=True,
)

for epoch in range(3):
    ds.set_epoch(epoch)
    for batch in loader:
        pass
```

* **Before this PR:** ~103.5 seconds per epoch.
* **After #8387:** ~65.8 seconds per epoch.
* **After this PR :** ~38.36s seconds per epoch.
**62.9% overall speedup**

### CODE REVIEW

**Comment by Yonghui-Lee** [comment]
The PR changes the shuffle results for a given seed, seems to cause test_iterable_dataset_shuffle_buffer_uses_multiple_input_shards to fail.

File: src/datasets/iterable_dataset.py [hunk 1]
```diff
@@ -1963,21 +1963,37 @@ def __iter__(self):
...
-                mem_buffer.append((key, pa_table))
-        # when we run out of examples, we shuffle the remaining examples in the buffer and yield them
-        rng.shuffle(mem_buffer)
-        yield from mem_buffer
+            last_key = key
+            tables.append(pa_table)
+            current_len += len(pa_table)
+
+            # Amortize shuffle cost by waiting until buffer is 2x full
+            if current_len >= 2 * self.buffer_size:
```
Comment (Yonghui-Lee) [inline src/datasets/iterable_dataset.py]: It kind of exceeds buffer size limit. Will it be better if we wait until buffer full, then keep half and yield half?

Comment (yuxin00j) [reply src/datasets/iterable_dataset.py]: Updated the logic to implement this.

**Comment by yuxin00j** [comment]
> The PR changes the shuffle results for a given seed, seems to cause test_iterable_dataset_shuffle_buffer_uses_multiple_input_shards to fail.

Updated the test.

**Review by lhoestq — COMMENTED** [review]
Nice improvement ! I think it would make sense to also update the `__iter__` method as well so both `__iter__` and `_iter_arrow` return examples in the same order - otherwise it can be confusing for users IMO

**Comment by yuxin00j** [comment]
> Nice improvement ! I think it would make sense to also update the `__iter__` method as well so both `__iter__` and `_iter_arrow` return examples in the same order - otherwise it can be confusing for users IMO

@lhoestq sure, I have updated the `__iter__` method.

**Review by lhoestq — APPROVED** [review]
lgtm ! I could even get a x150 speed boost with a local dataset !

**Comment by lhoestq** [comment]
This is really, really useful btw ! Would it be ok if I share this improvement online and mention you and the GCS team ?

edit: I couldn't wait lol: https://x.com/lhoestq/status/2096966828547277195?s=20

**Comment by yuxin00j** [comment]
> This is really, really useful btw ! Would it be ok if I share this improvement online and mention you and the GCS team ?
> 
> edit: I couldn't wait lol: https://x.com/lhoestq/status/2096966828547277195?s=20

@lhoestq Yes, thanks for the recognition!
